### PR TITLE
[DSE] Fix non-determinism due to address reuse

### LINKED ISSUE #84458: Optimization result is non-deterministic when compiling abc
Reduced test case: 
```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

define ptr @Vec_PtrAllocExact(i32 %0) {
  %2 = call ptr @malloc(i64 16)
  %3 = getelementptr i8, ptr %2, i64 4
  store i32 0, ptr %3, align 4
  store i32 %0, ptr %2, align 8
  %4 = getelementptr i8, ptr %2, i64 8
  store ptr null, ptr %4, align 8
  ret ptr %2
}

define internal void @Vec_PtrPush(ptr %0, ptr %1) {
  store i32 0, ptr %0, align 4
  ret void
}

declare i32 @Vec_PtrSize()

declare ptr @Abc_NtkObj()

declare ptr @Ptr_AbcDeriveOutputs()

define ptr @Ptr_AbcDeriveNtk() {
  %1 = call ptr @Vec_PtrAllocExact(i32 1)
  call void @Vec_PtrPush(ptr %1, ptr null)
  %2 = call ptr @Vec_PtrAllocExact(i32 0)
  store volatile ptr %2, ptr %1, align 8
  ret ptr null
}

declare ptr @malloc(i64)
```

Reproduction script:
```
#!/usr/bin/bash

bin/opt -O3 $1 -S -o diff1.ll
for i in $(seq 1 10);
do
    bin/opt -O3 $1 -S -o diff2.ll
    diff -q diff1.ll diff2.ll
    if [ $? -ne 0 ]; then
        exit 0
    fi
done

exit 1
```

Version1:
```
; ModuleID = 'reduced.ll'
source_filename = "reduced.ll"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

; Function Attrs: mustprogress nofree nounwind willreturn memory(write, argmem: none, inaccessiblemem: readwrite)
define noalias noundef ptr @Vec_PtrAllocExact(i32 %0) local_unnamed_addr #0 {
  %2 = tail call dereferenceable_or_null(16) ptr @malloc(i64 16)
  %3 = getelementptr i8, ptr %2, i64 4
  store i32 0, ptr %3, align 4
  store i32 %0, ptr %2, align 8
  %4 = getelementptr i8, ptr %2, i64 8
  store ptr null, ptr %4, align 8
  ret ptr %2
}

; Function Attrs: nofree nounwind memory(readwrite, argmem: none)
define noalias noundef ptr @Ptr_AbcDeriveNtk() local_unnamed_addr #1 {
  %1 = tail call dereferenceable_or_null(16) ptr @malloc(i64 16)
  %2 = getelementptr i8, ptr %1, i64 8
  store ptr null, ptr %2, align 8
  %calloc = tail call dereferenceable_or_null(16) ptr @calloc(i64 1, i64 16)
  store volatile ptr %calloc, ptr %1, align 8
  ret ptr null
}

; Function Attrs: mustprogress nofree nounwind willreturn allockind("alloc,uninitialized") allocsize(0) memory(inaccessiblemem: readwrite)
declare noalias noundef ptr @malloc(i64 noundef) local_unnamed_addr #2

; Function Attrs: nofree nounwind willreturn allockind("alloc,zeroed") allocsize(0,1) memory(inaccessiblemem: readwrite)
declare noalias noundef ptr @calloc(i64 noundef, i64 noundef) local_unnamed_addr #3

attributes #0 = { mustprogress nofree nounwind willreturn memory(write, argmem: none, inaccessiblemem: readwrite) }
attributes #1 = { nofree nounwind memory(readwrite, argmem: none) }
attributes #2 = { mustprogress nofree nounwind willreturn allockind("alloc,uninitialized") allocsize(0) memory(inaccessiblemem: readwrite) "alloc-family"="malloc" }
attributes #3 = { nofree nounwind willreturn allockind("alloc,zeroed") allocsize(0,1) memory(inaccessiblemem: readwrite) "alloc-family"="malloc" }
```

Version2:
```
; ModuleID = 'reduced.ll'
source_filename = "reduced.ll"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

; Function Attrs: mustprogress nofree nounwind willreturn memory(write, argmem: none, inaccessiblemem: readwrite)
define noalias noundef ptr @Vec_PtrAllocExact(i32 %0) local_unnamed_addr #0 {
  %2 = tail call dereferenceable_or_null(16) ptr @malloc(i64 16)
  %3 = getelementptr i8, ptr %2, i64 4
  store i32 0, ptr %3, align 4
  store i32 %0, ptr %2, align 8
  %4 = getelementptr i8, ptr %2, i64 8
  store ptr null, ptr %4, align 8
  ret ptr %2
}

; Function Attrs: nofree nounwind memory(readwrite, argmem: none)
define noalias noundef ptr @Ptr_AbcDeriveNtk() local_unnamed_addr #1 {
  %1 = tail call dereferenceable_or_null(16) ptr @malloc(i64 16)
  %2 = getelementptr i8, ptr %1, i64 4
  store i32 0, ptr %2, align 4
  %3 = getelementptr i8, ptr %1, i64 8
  store ptr null, ptr %3, align 8
  store i32 0, ptr %1, align 4
  %calloc = tail call dereferenceable_or_null(16) ptr @calloc(i64 1, i64 16)
  store volatile ptr %calloc, ptr %1, align 8
  ret ptr null
}

; Function Attrs: mustprogress nofree nounwind willreturn allockind("alloc,uninitialized") allocsize(0) memory(inaccessiblemem: readwrite)
declare noalias noundef ptr @malloc(i64 noundef) local_unnamed_addr #2

; Function Attrs: nofree nounwind willreturn allockind("alloc,zeroed") allocsize(0,1) memory(inaccessiblemem: readwrite)
declare noalias noundef ptr @calloc(i64 noundef, i64 noundef) local_unnamed_addr #3

attributes #0 = { mustprogress nofree nounwind willreturn memory(write, argmem: none, inaccessiblemem: readwrite) }
attributes #1 = { nofree nounwind memory(readwrite, argmem: none) }
attributes #2 = { mustprogress nofree nounwind willreturn allockind("alloc,uninitialized") allocsize(0) memory(inaccessiblemem: readwrite) "alloc-family"="malloc" }
attributes #3 = { nofree nounwind willreturn allockind("alloc,zeroed") allocsize(0,1) memory(inaccessiblemem: readwrite) "alloc-family"="malloc" }
```

I found this bug in my [benchmark](https://github.com/dtcxzyw/llvm-opt-benchmark/blob/main/bench/abc/original/bacPtrAbc.c.ll).
See also https://github.com/dtcxzyw/llvm-opt-benchmark/issues/334.

cc @nikic

### PR DESCRIPTION
**Pull request by nikic**
The malloc->calloc fold creates a new MemoryAccess, which may end of at the same address as a previously deleted access inside SkipStores.

To the most part, this is not a problem, because SkipStores is normally only used together with MemDefs. Neither the old malloc access nor the new calloc access will be part of MemDefs, so there is no problem here.

However, SkipStores is also used in one more place: In the main DSE loop, ToCheck entries are checked against it. Fix this by not using SkipStores here, and instead using a separate set to track deletions inside this loop. This way it is not affected by the calloc optimization that happens outside it.

This is all pretty ugly, but I haven't found another good way to fix it. Suggestions welcome.

No test case as I don't have a reliable DSE-only test-case for this.

Fixes https://github.com/llvm/llvm-project/issues/84458.

### CODE REVIEW

**Comment by nikic** [comment]
ping

**Comment by nikic** [comment]
ping

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 1]
```diff
@@ -2167,14 +2171,15 @@ static bool eliminateDeadStores(Function &F, AliasAnalysis &AA, MemorySSA &MSSA,
     unsigned PartialLimit = MemorySSAPartialStoreLimit;
     // Worklist of MemoryAccesses that may be killed by KillingDef.
     SmallSetVector<MemoryAccess *, 8> ToCheck;
+    SmallPtrSet<MemoryAccess *, 8> Deleted;
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2174]: Could you add a comment here explaining why this extra set is needed here?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 2]
```diff
@@ -2221,7 +2226,7 @@ static bool eliminateDeadStores(Function &F, AliasAnalysis &AA, MemorySSA &MSSA,
           continue;
         LLVM_DEBUG(dbgs() << "DSE: Remove Dead Store:\n  DEAD: " << *DeadI
                           << "\n  KILLER: " << *KillingI << '\n');
-        State.deleteDeadInstruction(DeadI);
+        State.deleteDeadInstruction(DeadI, &Deleted);
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2229]: Would it be possible to check if all uses in the loop pass Deleted, e.g. add an assert that checks if we add as many elements to `SkipStore` as we do to `Deleted` after the loop?

**Review by fhahn — COMMENTED** [review]
Thanks for the patch, I can't really think of a better solution at the moment.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp (hunk 2 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2234]: I've added an assertion -- does that match what you had in mind?

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:2234]: Thanks!

**Pull request merged**